batch.py

```diff
@@ -0,0 +1,324 @@
+
+"""
+Hail Batch workflow for the rare-variant association analysis, including:
+
+- get relevant variants around a gene and export genotypes as plink files
+- generate input files for association tests 
+- run association tests
+"""
+
+# import python modules # is there a specific order, rationale for grouping imports??
+import os
+import re
+
+import click
+import logging
+
+from cloudpathlib import AnyPath
+from cpg_utils import to_path
+from cpg_utils.hail_batch import (
+    dataset_path,
+    get_config,
+    init_batch,
+    output_path,
+    remote_tmpdir,
+)
+
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+from limix.qc import quantile_gaussianize
+from numpy import eye, ones
+from pandas_plink import read_plink1_bin
+from scipy.stats import shapiro
+
+# qvalue imports
+import scipy as sp
+from scipy import interpolate
+
+# make gene loc dict import
+from csv import DictReader
+
+import hail as hl
+import hailtop.batch as hb
+from hail.methods import export_plink
+
+# use logging to print statements, display at info level
+logging.basicConfig(
+    format="%(levelname)s:%(message)s", level=logging.INFO
+)  # consider removing some print statements?
+
+DEFAULT_JOINT_CALL_MT = dataset_path("mt/v7.mt")
+DEFAULT_ANNOTATION_HT = dataset_path(
+    "tob_wgs_vep/104/vep104.3_GRCh38.ht"
+)  # atm VEP only
+
+CELLREGMAP_IMAGE = (
+    "australia-southeast1-docker.pkg.dev/cpg-common/images/cellregmap:0.0.3"
+)
+
+# region SUBSET_VARIANTS
+
+
+def filter_variants(
+    mt_path: str,  # "mt/v7.mt"
+    samples: list[str],
+    output_mt_path: str,  # "tob_wgs_rv/densified_rv_only.mt"
+):
+    """Subset hail matrix table
+
+    Input:
+    joint call hail matrix table
+    set of samples for which we have scRNA-seq data
+
+    Output:
+    subsetted hail matrix table, containing only variants that:
+    1) are not ref-only, 2) biallelic, 3) meet QC filters, 4) are rare (MAF<5%)
+    """
+    # read hail matrix table object (WGS data)
+    init_batch()
+    mt = hl.read_matrix_table(mt_path)
+
+    # subset to relevant samples (samples we have scRNA-seq data for)
+    mt = mt.filter_cols(hl.set(samples).contains(mt.s))
+
+    # densify (can this be done at the end?)
+    mt = hl.experimental.densify(mt)
+
+    # filter out low quality variants and consider biallelic variants only (no multi-allelic, no ref-only)
+    mt = mt.filter_rows(  # check these filters!
+        (hl.len(hl.or_else(mt.filters, hl.empty_set(hl.tstr))) == 0)  # QC
+        & (hl.len(mt.alleles) == 2)  # remove hom-ref
+        & (mt.n_unsplit_alleles == 2)  # biallelic
+        & (hl.is_snp(mt.alleles[0], mt.alleles[1]))  # SNVs
+    )
+
+    # filter rare variants only (MAF < 5%)
+    mt = hl.variant_qc(mt)
+    mt = mt.filter_rows(
+        (mt.variant_qc.AF[1] < 0.05) & (mt.variant_qc.AF[1] > 0)
+        | (mt.variant_qc.AF[1] > 0.95) & (mt.variant_qc.AF[1] < 1)
+    )
+    mt = mt.checkpoint(output_mt_path, overwrite=True)
```

Comment:
You should only use checkpoint for temporary results, as `checkpoint` uses a faster codec that produces larger results. For permanent results, use `write` instead.

You'll have to reread the matrix table again afterwards though, as `write` doesn't return anything.

Reply:
i think in this case i was not returning anything anyways, so should be ok to just turn this checkpoint into write?